core/src/main/java/org/jdbi/v3/tweak/transactions/LocalTransactionHandler.java

```diff
@@ -165,36 +165,28 @@ public boolean isInTransaction(Handle handle)
     }
 
     @Override
-    public <ReturnType> ReturnType inTransaction(Handle handle, TransactionCallback<ReturnType> callback)
+    public <R, X extends Exception> R inTransaction(Handle handle,
+                                                    TransactionCallback<R, X> callback) throws X
     {
         final AtomicBoolean failed = new AtomicBoolean(false);
         TransactionStatus status = () -> failed.set(true);
-        final ReturnType returnValue;
+        final R returnValue;
         try {
             handle.begin();
             returnValue = callback.inTransaction(handle, status);
             if (!failed.get()) {
                 handle.commit();
             }
         }
-        catch (RuntimeException e) {
-            try {
-                handle.rollback();
-            } catch (Exception rollback) {
-                e.addSuppressed(rollback);
-            }
-            throw e;
-        }
         catch (Exception e) {
             try {
                 handle.rollback();
             } catch (Exception rollback) {
                 e.addSuppressed(rollback);
             }
-            throw new TransactionFailedException("Transaction failed do to exception being thrown " +
-                                                 "from within the callback. See cause " +
-                                                 "for the original exception.", e);
+            throw (X) e;
```

Comment:
My reasoning is that the try can only throw a `RuntimeException` or `X`. So the cast is safe from a runtime perspective. At compile time `X` is erased to just `Exception`.

Plus there's the whole part about how checked exceptions are a purely compile time construct--the runtime makes no distinction and any method can legally throw any Throwable they like.